File: OpenHome/Net/Bindings/Cs/ControlPoint/Proxies/CpAvOpenhomeOrgWebProcessorConfig1.cs
using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using OpenHome.Net.Core;
using OpenHome.Net.ControlPoint;

namespace OpenHome.Net.ControlPoint.Proxies
{
    public interface ICpProxyAvOpenhomeOrgWebProcessorConfig1 : ICpProxy, IDisposable
    {
        void SyncGetProcessorConfig(out String aProcessorConfig);
        void BeginGetProcessorConfig(CpProxy.CallbackAsyncComplete aCallback);
        void EndGetProcessorConfig(IntPtr aAsyncHandle, out String aProcessorConfig);
        void SyncSetProcessorConfig(String aProcessorConfig);
        void BeginSetProcessorConfig(String aProcessorConfig, CpProxy.CallbackAsyncComplete aCallback);
        void EndSetProcessorConfig(IntPtr aAsyncHandle);
        void SetPropertyAliveChanged(System.Action aAliveChanged);
        bool PropertyAlive();
        void SetPropertyProcessorConfigChanged(System.Action aProcessorConfigChanged);
        String PropertyProcessorConfig();
    }

    internal class SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgWebProcessorConfig1 iService;
        private String iProcessorConfig;

        public SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1 aProxy)
        {
            iService = aProxy;
        }
        public String ProcessorConfig()
        {
            return iProcessorConfig;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndGetProcessorConfig(aAsyncHandle, out iProcessorConfig);
        }
    };

    internal class SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgWebProcessorConfig1 iService;

        public SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1 aProxy)
        {
            iService = aProxy;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndSetProcessorConfig(aAsyncHandle);
        }
    };

    /// <summary>
    /// Proxy for the av.openhome.org:WebProcessorConfig:1 UPnP service
    /// </summary>
    public class CpProxyAvOpenhomeOrgWebProcessorConfig1 : CpProxy, IDisposable, ICpProxyAvOpenhomeOrgWebProcessorConfig1
    {
        private OpenHome.Net.Core.Action iActionGetProcessorConfig;
        private OpenHome.Net.Core.Action iActionSetProcessorConfig;
        private PropertyBool iAlive;
        private PropertyString iProcessorConfig;
        private System.Action iAliveChanged;
        private System.Action iProcessorConfigChanged;
        private Mutex iPropertyLock;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <remarks>Use CpProxy::[Un]Subscribe() to enable/disable querying of state variable and reporting of their changes.</remarks>
        /// <param name="aDevice">The device to use</param>
        public CpProxyAvOpenhomeOrgWebProcessorConfig1(CpDevice aDevice)
            : base("av-openhome-org", "WebProcessorConfig", 1, aDevice)
        {
            OpenHome.Net.Core.Parameter param;
            List<String> allowedValues = new List<String>();

            iActionGetProcessorConfig = new OpenHome.Net.Core.Action("GetProcessorConfig");
            param = new ParameterString("ProcessorConfig", allowedValues);
            iActionGetProcessorConfig.AddOutputParameter(param);

            iActionSetProcessorConfig = new OpenHome.Net.Core.Action("SetProcessorConfig");
            param = new ParameterString("ProcessorConfig", allowedValues);
            iActionSetProcessorConfig.AddInputParameter(param);

            iAlive = new PropertyBool("Alive", AlivePropertyChanged);
            AddProperty(iAlive);
            iProcessorConfig = new PropertyString("ProcessorConfig", ProcessorConfigPropertyChanged);
            AddProperty(iProcessorConfig);
            
            iPropertyLock = new Mutex();
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aProcessorConfig"></param>
        public void SyncGetProcessorConfig(out String aProcessorConfig)
        {
            SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 sync = new SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(this);
            BeginGetProcessorConfig(sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
            aProcessorConfig = sync.ProcessorConfig();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndGetProcessorConfig().</remarks>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginGetProcessorConfig(CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionGetProcessorConfig, aCallback);
            int outIndex = 0;
            invocation.AddOutput(new ArgumentString((ParameterString)iActionGetProcessorConfig.OutputParameter(outIndex++)));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        /// <param name="aProcessorConfig"></param>
        public void EndGetProcessorConfig(IntPtr aAsyncHandle, out String aProcessorConfig)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
            uint index = 0;
            aProcessorConfig = Invocation.OutputString(aAsyncHandle, index++);
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aProcessorConfig"></param>
        public void SyncSetProcessorConfig(String aProcessorConfig)
        {
            SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 sync = new SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(this);
            BeginSetProcessorConfig(aProcessorConfig, sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndSetProcessorConfig().</remarks>
        /// <param name="aProcessorConfig"></param>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginSetProcessorConfig(String aProcessorConfig, CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionSetProcessorConfig, aCallback);
            int inIndex = 0;
            invocation.AddInput(new ArgumentString((ParameterString)iActionSetProcessorConfig.InputParameter(inIndex++), aProcessorConfig));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        public void EndSetProcessorConfig(IntPtr aAsyncHandle)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the Alive state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgWebProcessorConfig1 instance will not overlap.</remarks>
        /// <param name="aAliveChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyAliveChanged(System.Action aAliveChanged)
        {
            lock (iPropertyLock)
            {
                iAliveChanged = aAliveChanged;
            }
        }

        private void AlivePropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iAliveChanged);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the ProcessorConfig state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgWebProcessorConfig1 instance will not overlap.</remarks>
        /// <param name="aProcessorConfigChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyProcessorConfigChanged(System.Action aProcessorConfigChanged)
        {
            lock (iPropertyLock)
            {
                iProcessorConfigChanged = aProcessorConfigChanged;
            }
        }

        private void ProcessorConfigPropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iProcessorConfigChanged);
            }
        }

        /// <summary>
        /// Query the value of the Alive property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Alive property</returns>
        public bool PropertyAlive()
        {
            PropertyReadLock();
            bool val;
            try
            {
                val = iAlive.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Query the value of the ProcessorConfig property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the ProcessorConfig property</returns>
        public String PropertyProcessorConfig()
        {
            PropertyReadLock();
            String val;
            try
            {
                val = iProcessorConfig.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                    return;
                DisposeProxy();
                iHandle = IntPtr.Zero;
            }
            iActionGetProcessorConfig.Dispose();
            iActionSetProcessorConfig.Dispose();
            iAlive.Dispose();
            iProcessorConfig.Dispose();
        }
    }
}

File: OpenHome/Net/ControlPoint/Proxies/CpAvOpenhomeOrgWebProcessorConfig1.cpp
#include "CpAvOpenhomeOrgWebProcessorConfig1.h"
#include <OpenHome/Net/Core/CpProxy.h>
#include <OpenHome/Net/Private/CpiService.h>
#include <OpenHome/Private/Thread.h>
#include <OpenHome/Net/Private/AsyncPrivate.h>
#include <OpenHome/Net/Core/CpDevice.h>
#include <OpenHome/Net/Private/Error.h>
#include <OpenHome/Net/Private/CpiDevice.h>

namespace OpenHome {
namespace Net {

class SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 : public SyncProxyAction
{
public:
    SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1& aProxy, Brh& aProcessorConfig);
    virtual void CompleteRequest(IAsync& aAsync);
private:
    CpProxyAvOpenhomeOrgWebProcessorConfig1& iService;
    Brh& iProcessorConfig;
};

class SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 : public SyncProxyAction
{
public:
    SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1& aProxy);
    virtual void CompleteRequest(IAsync& aAsync);
private:
    CpProxyAvOpenhomeOrgWebProcessorConfig1& iService;
};

} // namespace Net
} // namespace OpenHome


using namespace OpenHome;
using namespace OpenHome::Net;


// SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1

SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1::SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1& aProxy, Brh& aProcessorConfig)
    : iService(aProxy)
    , iProcessorConfig(aProcessorConfig)
{
}

void SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1::CompleteRequest(IAsync& aAsync)
{
    iService.EndGetProcessorConfig(aAsync, iProcessorConfig);
}

// SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1

SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1::SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1(CpProxyAvOpenhomeOrgWebProcessorConfig1& aProxy)
    : iService(aProxy)
{
}

void SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1::CompleteRequest(IAsync& aAsync)
{
    iService.EndSetProcessorConfig(aAsync);
}


// CpProxyAvOpenhomeOrgWebProcessorConfig1

CpProxyAvOpenhomeOrgWebProcessorConfig1::CpProxyAvOpenhomeOrgWebProcessorConfig1(CpDevice& aDevice)
    : iCpProxy("av-openhome-org", "WebProcessorConfig", 1, aDevice.Device())
{
    OpenHome::Net::Parameter* param;

    iActionGetProcessorConfig = new Action("GetProcessorConfig");
    param = new OpenHome::Net::ParameterString("ProcessorConfig");
    iActionGetProcessorConfig->AddOutputParameter(param);

    iActionSetProcessorConfig = new Action("SetProcessorConfig");
    param = new OpenHome::Net::ParameterString("ProcessorConfig");
    iActionSetProcessorConfig->AddInputParameter(param);

    Functor functor;
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgWebProcessorConfig1::AlivePropertyChanged);
    iAlive = new PropertyBool("Alive", functor);
    AddProperty(iAlive);
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgWebProcessorConfig1::ProcessorConfigPropertyChanged);
    iProcessorConfig = new PropertyString("ProcessorConfig", functor);
    AddProperty(iProcessorConfig);
}

CpProxyAvOpenhomeOrgWebProcessorConfig1::~CpProxyAvOpenhomeOrgWebProcessorConfig1()
{
    DestroyService();
    delete iActionGetProcessorConfig;
    delete iActionSetProcessorConfig;
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SyncGetProcessorConfig(Brh& aProcessorConfig)
{
    SyncGetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 sync(*this, aProcessorConfig);
    BeginGetProcessorConfig(sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::BeginGetProcessorConfig(FunctorAsync& aFunctor)
{
    Invocation* invocation = iCpProxy.GetService().Invocation(*iActionGetProcessorConfig, aFunctor);
    TUint outIndex = 0;
    const Action::VectorParameters& outParams = iActionGetProcessorConfig->OutputParameters();
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    iCpProxy.GetInvocable().InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::EndGetProcessorConfig(IAsync& aAsync, Brh& aProcessorConfig)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("GetProcessorConfig"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
    TUint index = 0;
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aProcessorConfig);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SyncSetProcessorConfig(const Brx& aProcessorConfig)
{
    SyncSetProcessorConfigAvOpenhomeOrgWebProcessorConfig1 sync(*this);
    BeginSetProcessorConfig(aProcessorConfig, sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::BeginSetProcessorConfig(const Brx& aProcessorConfig, FunctorAsync& aFunctor)
{
    Invocation* invocation = iCpProxy.GetService().Invocation(*iActionSetProcessorConfig, aFunctor);
    TUint inIndex = 0;
    const Action::VectorParameters& inParams = iActionSetProcessorConfig->InputParameters();
    invocation->AddInput(new ArgumentString(*inParams[inIndex++], aProcessorConfig));
    iCpProxy.GetInvocable().InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::EndSetProcessorConfig(IAsync& aAsync)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("SetProcessorConfig"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SetPropertyAliveChanged(Functor& aFunctor)
{
    iCpProxy.GetLock().Wait();
    iAliveChanged = aFunctor;
    iCpProxy.GetLock().Signal();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SetPropertyProcessorConfigChanged(Functor& aFunctor)
{
    iCpProxy.GetLock().Wait();
    iProcessorConfigChanged = aFunctor;
    iCpProxy.GetLock().Signal();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::PropertyAlive(TBool& aAlive) const
{
    AutoMutex a(iCpProxy.PropertyReadLock());
    if (iCpProxy.GetSubscriptionStatus() != CpProxy::eSubscribed) {
        THROW(ProxyNotSubscribed);
    }
    aAlive = iAlive->Value();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::PropertyProcessorConfig(Brhz& aProcessorConfig) const
{
    AutoMutex a(iCpProxy.PropertyReadLock());
    if (iCpProxy.GetSubscriptionStatus() != CpProxy::eSubscribed) {
        THROW(ProxyNotSubscribed);
    }
    aProcessorConfig.Set(iProcessorConfig->Value());
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::AlivePropertyChanged()
{
    ReportEvent(iAliveChanged);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::ProcessorConfigPropertyChanged()
{
    ReportEvent(iProcessorConfigChanged);
}


void CpProxyAvOpenhomeOrgWebProcessorConfig1::Subscribe()
{
  iCpProxy.Subscribe();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::Unsubscribe()
{
 iCpProxy.Unsubscribe();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SetPropertyChanged(Functor& aFunctor)
{
  iCpProxy.SetPropertyChanged(aFunctor);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::SetPropertyInitialEvent(Functor& aFunctor)
{
  iCpProxy.SetPropertyInitialEvent(aFunctor);
}
void CpProxyAvOpenhomeOrgWebProcessorConfig1::AddProperty(Property* aProperty)
{
  iCpProxy.AddProperty(aProperty);
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::DestroyService()
{
  iCpProxy.DestroyService();
}

void CpProxyAvOpenhomeOrgWebProcessorConfig1::ReportEvent(Functor aFunctor)
{
  iCpProxy.ReportEvent(aFunctor);
}

TUint CpProxyAvOpenhomeOrgWebProcessorConfig1::Version() const
{
  return iCpProxy.Version();
}


